x/mint/module/abci.go

```diff
@@ -2,80 +2,180 @@ package mint
+			types.ErrNegativeCirculatingSupply,
+			"total supply %s, locked supply %s",
+			totalSupply.String(),
+			lockedSupply.String(),
+		)
+	}
+	fmt.Println("Total supply", totalSupply)
+	fmt.Println("Locked supply", lockedSupply)
+	fmt.Println("Circulating supply", circulatingSupply)
+	fmt.Println("FEmissionNumerator", params.FEmissionNumerator)
+	fmt.Println("FEmissionDenominator", params.FEmissionDenominator)
+	targetRewardEmissionPerUnitStakedToken,
+		err := keeper.GetTargetRewardEmissionPerUnitStakedToken(
+		params.FEmissionNumerator,
+		params.FEmissionDenominator,
+		ecosystemBalance,
+		networkStaked,
+		circulatingSupply,
+		totalSupply,
+	)
+	fmt.Println("Target reward emission per unit staked token", targetRewardEmissionPerUnitStakedToken)
+	if err != nil {
+		return math.Int{}, math.LegacyDec{}, err
+	}
+	smoothingDegree := keeper.GetSmoothingFactorPerTimestep(
+		ctx,
+		k,
+		params.OneMonthSmoothingDegreeNumerator,
+		params.OneMonthSmoothingDegreeDenominator,
+		params.EmissionCalibrationsTimestepPerMonth,
+	)
+	fmt.Println("Smoothing degree", smoothingDegree)
+	previousRewardEmissionPerUnitStakedToken, err := k.PreviousRewardEmissionPerUnitStakedToken.Get(ctx)
+	if err != nil {
+		return math.Int{}, math.LegacyDec{}, err
+	}
+	fmt.Println("Previous reward emissions per unit staked token", previousRewardEmissionPerUnitStakedToken)
+	emissionPerUnitStakedToken = keeper.GetRewardEmissionPerUnitStakedToken(
+		targetRewardEmissionPerUnitStakedToken,
+		smoothingDegree,
+		previousRewardEmissionPerUnitStakedToken,
+	)
+	fmt.Println("E_i", emissionPerUnitStakedToken)
+	emissionPerTimestep = keeper.GetTotalEmissionPerTimestep(emissionPerUnitStakedToken, networkStaked)
+	return emissionPerTimestep, emissionPerUnitStakedToken, nil
+}
+
 func BeginBlocker(ctx context.Context, k keeper.Keeper) error {
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
 
-	minter, err := k.Minter.Get(ctx)
+	params, err := k.Params.Get(ctx)
 	if err != nil {
 		return err
 	}
-	params, err := k.Params.Get(ctx)
+	// Get the balance of the "ecosystem" module account
+	ecosystemBalance, err := k.GetEcosystemBalance(ctx, params.MintDenom)
+	fmt.Println("Ecosystem balance", ecosystemBalance)
 	if err != nil {
 		return err
 	}
 
-	inflationBeforeUpdates := minter.Inflation
-	totalCirculatingSupply := k.GetSupply(ctx).Amount
-
-	currentBlockProvision := params.CurrentBlockProvision
-
-	// Calculate the new total supply if coins were minted this block
-	newTotalCirculatingSupply := totalCirculatingSupply.Add(cosmosMath.Int(currentBlockProvision))
-
-	// Only mint new coins if new total supply would not exceed max supply
-	if newTotalCirculatingSupply.LTE(cosmosMath.Int(params.MaxSupply)) {
-		mintedCoin := sdk.NewCoin(params.MintDenom, cosmosMath.Int(currentBlockProvision))
-		mintedCoins := sdk.NewCoins(mintedCoin)
+	// find out if we need to update the block emissions rate
+	emissionRateUpdateCadence := params.BlocksPerMonth / params.EmissionCalibrationsTimestepPerMonth
```

Comment:
bold! a raw division. does this need to be protected like all of the other divisions?

Reply:
I think it's fine, truncation is fine with me in general for this

Reply:
we care about other divisions where we're dealing with decimal-like types or financial coin balances, but this is just to figure out how often to do something